### PR TITLE
[WIP] jax 1d

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This in progress PR includes the jax implementation of the 1d scattering transform. Currently the tests do not pass. I am going through in a google collab notebook and comparing the numpy and jax functions. 

This PR is part of my work for the Flatiron internship.

### CODE REVIEW

File: tests/scattering1d/view_filters.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from kymatio import Scattering1D
```
Comment (eickenberg) [inline tests/scattering1d/view_filters.py:1]: Why is this file in the tests folder?

File: tests/scattering1d/test_jax_scattering1d.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import jax.numpy as np
+from jax import device_put
+import io
+
+backends = []
+
+from kymatio.scattering1d.backend.jax_backend import backend
+backends.append(backend)
+
+class TestScattering1DJax:
```
Comment (eickenberg) [inline tests/scattering1d/test_jax_scattering1d.py:14]: unnecessesary

File: tests/scattering1d/test_jnp_vs_np.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
+# -*- coding: utf-8 -*-
+"""jaxvsnumpy.ipynb
+
+Automatically generated by Colaboratory.
```
Comment (eickenberg) [inline tests/scattering1d/test_jnp_vs_np.py:4]: ❓

File: tests/scattering1d/test_jnp_vs_np.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+def np_pad(x, pad_left, pad_right):
+  res = np.pad(x, ((0, 0), (0, 0), (pad_left, pad_right),), mode='reflect')
+  return res
+
+def jnp_pad(x, pad_left, pad_right):
+  res = jnp.pad(x, ((0, 0), (0, 0), (pad_left, pad_right),), mode='reflect')
+  return res
+
+class TestNumpyJaxMultiplication:
+    def test_mulitplication(self):
```
Comment (eickenberg) [inline tests/scattering1d/test_jnp_vs_np.py:36]: typo

File: tests/scattering1d/test_jnp_vs_np.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        f = np.real(c)
+        assert jnp.allclose(f, l)
+        
+    def test_subsample_fourier(self):
+        #subsample fourier
+        a = np.random.rand(4, 1, 1024)
+        b = device_put(jnp.asarray(a))
+        for i in [1, 2, 4]:
+            c = subsample_fourier(a, i)
+            d = subsample_fourier(a, i)
```
Comment (eickenberg) [inline tests/scattering1d/test_jnp_vs_np.py:201]: what does this test? looks like it is calling 2x the same thing

File: tests/scattering1d/test_jnp_vs_np.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            assert jnp.allclose(c, d)
+    
+    def test_jnp_subsample_fourier(self):       
+        #subsample fourier jnp
+        a = np.random.rand(4, 1, 1024)
+        b = device_put(jnp.asarray(a))
+        for i in [1, 2, 4]:
+            c = subsample_fourier(a, i)
+            d = subsample_fourier(a, i)
+            assert jnp.allclose(c, d)
```
Comment (eickenberg) [inline tests/scattering1d/test_jnp_vs_np.py:211]: looks like c and d are identical, is this intended?

File: tests/scattering1d/test_jnp_vs_np.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+    def test_pad_unpad(self):       
+        #pad
+        a = np.random.rand(4, 1, 512)
+        b = np_pad(a, 256, 256)
+        c = device_put(jnp.asarray(a))
+        d = jnp_pad(c, 256, 256)
+        assert jnp.allclose(b, d)
+        b = b[..., 4: 12]
```
Comment (eickenberg) [inline tests/scattering1d/test_jnp_vs_np.py:220]: remove space. same for row below

File: tests/scattering1d/test_jnp_vs_np.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    
+    def test_jnp_pad_unpad(self):       
+        #pad jnp
+        a = device_put(random.normal(key, (4, 1, 512)))
+        b = jnp_pad(a, 256, 256)
+        c = np.asarray(a)
+        d = np_pad(c, 256, 256)
+        assert jnp.allclose(b, d)
+        b = b[..., 4: 12]
+        d = d[..., 4: 12]
```
Comment (eickenberg) [inline tests/scattering1d/test_jnp_vs_np.py:232]: remove space. same for row above

**Comment by janden** [comment]
This looks like a mix of a whole set of changes, not just Jax. Is this branching off of the rfft branch (in #573)? Could you rebase this off of `dev` so only the Jax commits show up? If not, reviewing will have to wait until #573 is merged.

**Comment by janden** [comment]
I'm getting test failures locally:
```
=================================================== test session starts ====================================================
platform linux -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /shared/home/janden/projects/kymatio
plugins: cov-2.9.0
collected 56 items

tests/scattering1d/test_filters_scattering1d.py .........                                                            [ 16%]
tests/scattering1d/test_jax_scattering1d.py .                                                                        [ 17%]
tests/scattering1d/test_jnp_vs_np.py ......FFFF.........                                                             [ 51%]
tests/scattering1d/test_numpy_scattering1d.py .                                                                      [ 53%]
tests/scattering1d/test_tensorflow_scattering1d.py .                                                                 [ 55%]
tests/scattering1d/test_torch_scattering1d.py ......................                                                 [ 94%]
tests/scattering1d/test_utils_scattering1d.py ...                                                                    [100%]

========================================================= FAILURES =========================================================
_________________________________________________________ test_fft _________________________________________________________

    def test_fft():
        #fft
        real_filter = np.random.rand(1024)
        complex_input = np.random.rand(4, 1, 1024) + 1j * np.random.rand(4, 1, 1024)

        numpy_result = np.fft.fft(complex_input * real_filter)

        real_filter = device_put(jnp.asarray(real_filter))
        complex_input = device_put(jnp.asarray(complex_input))

        jax_result = jnp.fft.fft(complex_input * real_filter)
>       assert np.allclose(jax_result, numpy_result, atol=1e-6, rtol=1e-4)
E       assert False
E        +  where False = <function allclose at 0x7f12bd733a60>(DeviceArray([[[255.52388   +2.5527682e+02j,   0.9501991 +4.5342102e+00j,\n                14.7885    +4.4190798e+00j, ....1.0180717 -5.8162346e+00j,  -3.1163976 +3.4822960e+00j,\n                -5.9054456 +1.5221304e+01j]]], dtype=complex64), array([[[255.52386992+2.55276812e+02j,   0.95019845+4.53421088e+00j,\n          14.78849949+4.41907980e+00j, ...,\n     ...0j, ...,\n          -1.0180956 -5.81623633e+00j,  -3.11639267+3.48238648e+00j,\n          -5.90549594+1.52213110e+01j]]]), atol=1e-06, rtol=0.0001)
E        +    where <function allclose at 0x7f12bd733a60> = np.allclose

tests/scattering1d/test_jnp_vs_np.py:100: AssertionError
_______________________________________________________ test_jnp_fft _______________________________________________________

    def test_jnp_fft():
        #fft jnp
        real_filter = device_put(random.normal(key, (1024,)))
        complex_input = device_put(random.normal(key, (4, 1, 1024)) + 1j * random.normal(key, (4, 1, 1024)))

        jax_result = jnp.fft.fft(complex_input * real_filter)

        real_filter = np.asarray(real_filter)
        complex_input = np.asarray(complex_input)

        numpy_result = np.fft.fft(complex_input * real_filter)

>       assert jnp.allclose(numpy_result, jax_result, atol=1e-6, rtol=1e-4)
E       assert DeviceArray(False, dtype=bool)
E        +  where DeviceArray(False, dtype=bool) = <function allclose at 0x7f12b2e62dc0>(array([[[  5.98983673 +5.98983673j, -46.70631379 -4.87641587j,\n           7.71041395+35.87813789j, ...,  10.82137953-1...7338833-29.06419958j, ...,  -3.66540608+23.42813572j,\n         -29.06419958-13.57338833j, -51.25893607-10.82506061j]]]), DeviceArray([[[  5.9898376 +5.9898376j, -46.70631   -4.876417j ,\n                 7.7104144+35.878143j , ...,  10.8213...-3.6655445+23.427938j ,\n               -29.064018 -13.573335j , -51.25883  -10.824964j ]]],            dtype=complex64), atol=1e-06, rtol=0.0001)
E        +    where <function allclose at 0x7f12b2e62dc0> = jnp.allclose

tests/scattering1d/test_jnp_vs_np.py:114: AssertionError
________________________________________________________ test_rfft _________________________________________________________

    def test_rfft():
        #rfft
        x = np.random.rand(4, 1, 1024)
        numpy_result = np.fft.fft(x)

        x = device_put(jnp.asarray(x))
        jax_result = jnp.fft.fft(x)

>       assert np.allclose(jax_result, numpy_result, atol=1e-6, rtol=1e-4)
E       assert False
E        +  where False = <function allclose at 0x7f12bd733a60>(DeviceArray([[[ 5.17765381e+02+0.j        , -3.51818085e-01-1.0428596j ,\n               -6.16341352e-01-2.2897239j , .....00402565e+01-9.258729j  ,  4.08682632e+00+6.988961j  ,\n               -6.02400684e+00+6.4724164j ]]], dtype=complex64), array([[[ 5.17765417e+02+0.j        , -3.51818376e-01-1.04285927j,\n         -6.16341421e-01-2.28972346j, ..., -3.78746...00-6.98896118j, ..., -1.00402721e+01-9.25874196j,\n          4.08679967e+00+6.98896118j, -6.02400541e+00+6.472408j  ]]]), atol=1e-06, rtol=0.0001)
E        +    where <function allclose at 0x7f12bd733a60> = np.allclose

tests/scattering1d/test_jnp_vs_np.py:124: AssertionError
______________________________________________________ test_jnp_rfft _______________________________________________________

    def test_jnp_rfft():
        #rfft jnp
        x = device_put(random.normal(key, (4, 1, 1024)))
        jax_result = jnp.fft.fft(x)

        x = np.asarray(x)
        numpy_result = np.fft.fft(x)

>       assert jnp.allclose(numpy_result, jax_result, atol=1e-6, rtol=1e-4)
E       assert DeviceArray(False, dtype=bool)
E        +  where DeviceArray(False, dtype=bool) = <function allclose at 0x7f12b2e62dc0>(array([[[ 31.99175451 +0.j        ,  36.96782752-29.86428754j,\n         -18.18336846+47.29498126j, ...,  27.40941669-1...3299884-18.1372982j , ..., -41.75252096+19.30701284j,\n         -21.33299884+18.1372982j ,  14.46780365-10.51116762j]]]), DeviceArray([[[ 31.991749  +0.j       ,  36.96783  -29.864288j ,\n               -18.183369 +47.294975j , ...,  27.4092...41.75238  +19.306877j ,\n               -21.332912 +18.137201j ,  14.467833 -10.511034j ]]],            dtype=complex64), atol=1e-06, rtol=0.0001)
E        +    where <function allclose at 0x7f12b2e62dc0> = jnp.allclose

tests/scattering1d/test_jnp_vs_np.py:134: AssertionError
===================================================== warnings summary =====================================================
/home/janden/venv/kymatio/lib/python3.8/site-packages/jax/lib/xla_bridge.py:125
  /home/janden/venv/kymatio/lib/python3.8/site-packages/jax/lib/xla_bridge.py:125: UserWarning: No GPU/TPU found, falling back to CPU.
    warnings.warn('No GPU/TPU found, falling back to CPU.')

/home/janden/venv/kymatio/lib/python3.8/site-packages/tensorflow/python/pywrap_tensorflow_internal.py:15
  /home/janden/venv/kymatio/lib/python3.8/site-packages/tensorflow/python/pywrap_tensorflow_internal.py:15: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/scattering1d/test_tensorflow_scattering1d.py::TestScattering1DTensorFlow::test_Scattering1D[backend]
  /home/janden/venv/kymatio/lib/python3.8/site-packages/tensorflow/python/training/tracking/data_structures.py:718: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    if not isinstance(wrapped_dict, collections.Mapping):

tests/scattering1d/test_torch_scattering1d.py: 20 tests with warnings
  /shared/home/janden/projects/kymatio/kymatio/scattering1d/frontend/torch_frontend.py:93: DeprecationWarning: The vectorize option is deprecated and will be removed in version 0.3. Please set out_type='list' for equivalent functionality.
    warnings.warn("The vectorize option is deprecated and will be "

tests/scattering1d/test_utils_scattering1d.py::test_scattering1d_frontend
  /shared/home/janden/projects/kymatio/kymatio/frontend/entry.py:21: DeprecationWarning: Torch frontend is currently the default, but NumPy will become the default in the next version.
    warnings.warn("Torch frontend is currently the default, but NumPy will become the default in the next"

tests/scattering1d/test_utils_scattering1d.py::test_scattering1d_frontend
  /shared/home/janden/projects/kymatio/kymatio/scattering1d/filter_bank.py:309: UserWarning: Signal support is too small to avoid border effects
    warnings.warn('Signal support is too small to avoid border effects')

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================================= short test summary info ==================================================
FAILED tests/scattering1d/test_jnp_vs_np.py::test_fft - assert False
FAILED tests/scattering1d/test_jnp_vs_np.py::test_jnp_fft - assert DeviceArray(False, dtype=bool)
FAILED tests/scattering1d/test_jnp_vs_np.py::test_rfft - assert False
FAILED tests/scattering1d/test_jnp_vs_np.py::test_jnp_rfft - assert DeviceArray(False, dtype=bool)
======================================== 4 failed, 52 passed, 25 warnings in 18.17s ========================================
```
We also need to make sure jax and jaxlib are added to the test dependencies so that the tests can run on the CI.

**Comment by janden** [comment]
Aside from the import statement at the top, does the frontend and backend codes differ fundamentally from the NumPy codes? If I understand correctly, it's a stated goal of Jax to have their API be as compatible to NumPy as possible. For the parts that we're using, it certainly seems like it. If that's the case, is there some way we can take advantage of that to avoid duplicating the code?

One way would be to add some flag in the NumPy backend that selects whether we should import the regular NumPy or the Jax one. Alternatively, we can make `np` an argument to the backend functions and fix it when assigning them to the `NamedTuple`. (One thing that doesn't work is importing the backend and replacing the `np` symbol in there since that would break the regular NumPy backend.)

Maybe I'm overthinking this, but we are getting into a place where we have a significant amount of code to maintain with each “small” change since it has to be replicated across frontends and backends. Deduplicating the NumPy and Jax codes would alleviate this a little bit.

**Comment by MuawizChaudhary** [comment]
So some of the fft tests fail, but i can set the tolerances such that they do not fail. However, the tolerance is pretty high (its like an absolute tolerance of 1e-4 or 1e-5).

I'll try to look into not replicating the two codes. I think there are some differences that could make it impossible, but lemme see.

**Comment by janden** [comment]
> So some of the fft tests fail, but i can set the tolerances such that they do not fail. 

Well, ideally we'd like the tests to pass, so let's adjust the tolerances since it doesn't look like Jax will be fixing their FFT very soon…

**Comment by janden** [comment]
> Maybe I'm overthinking this, but we are getting into a place where we have a significant amount of code to maintain with each “small” change since it has to be replicated across frontends and backends. Deduplicating the NumPy and Jax codes would alleviate this a little bit.

Since the above idea has now been implemented in `dev`, I suggest we close this PR (and its sister PRs #621, #622) and reimplement this by inheriting from the new `NumpyBackend` class.

**Comment by edouardoyallon** [comment]
sounds good

**Comment by eickenberg** [comment]
That sounds like the right way of doing it
